tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/telemetryMetrics/util/InMemoryMetricReader.java

```diff
@@ -71,61 +87,139 @@ public CompletableResultCode shutdown() {
+                .collect(Collectors.toList());
+        assertThat("Found more than one data point for metric: " + id, data, hasSize(lessThan(2)));
 
-        Optional<LongPointData> gague = allMetrics.stream()
-                .filter(
-                        md -> md.getName().equals(id.name))
-                .flatMap(md -> md.getLongGaugeData().getPoints().stream())
-                .filter(point -> id.attributes.asMap().keySet().stream()
-                        .allMatch(key -> point.getAttributes().asMap().containsKey(key)
-                                && id.attributes.asMap().get(key)
-                                        .equals(point.getAttributes().asMap().get(key))))
-                // feeding the points into Long.compare in reverse order will return the largest first.
-                .sorted((pointOne, pointTwo) -> Long.compare(pointTwo.getEpochNanos(), pointOne.getEpochNanos()))
-                .findFirst();
-
-        return gague;
+        return data.isEmpty() ? Optional.empty() : Optional.of(data.get(0));
     }
 
-    public String getUnit(String metricName) {
-        try {
-            Collection<MetricData> allMetrics = collectionRegistration.collectAllMetrics();
-            Optional<MetricData> mathcingData = allMetrics.stream()
-                    .filter(
-                            md -> md.getName().equals(metricName))
-                    .findAny();
-
-            return mathcingData.get().getUnit();
-        } catch (NoSuchElementException e) {
-            // If we didn't find anything throwing an exception to fail the test is reasonable
-            throw new RuntimeException("Found no results for " + metricName);
+    private static <T extends PointData> Optional<T> getSumPointData(SumData<T> sumData, TelemetryMetricID id) {
+        assertEquals(sumData.getAggregationTemporality(), CUMULATIVE, "Wrong temporality type for metric " + id.name);
+
+        List<T> data = sumData.getPoints().stream()
+                .filter(hasAllAttributes(id.attributes))
+                .collect(Collectors.toList());
+        assertThat("Found more than one data point for metric: " + id, data, hasSize(lessThan(2)));
+
+        return data.isEmpty() ? Optional.empty() : Optional.of(data.get(0));
+    }
+
+    /**
+     * Returns a predicate which checks whether a {@code PointData} contains all of the given {@code Attributes}.
+     *
+     * @param attributes
+     *            the attributes to check for
+     * @return a predicate which returns {@code true} if {@link PointData#getAttributes()} returns a superset of
+     *         {@code attributes}, otherwise {@code false}
+     */
+    private static Predicate<PointData> hasAllAttributes(Attributes attributes) {
+        return pointData -> {
+            for (Entry<AttributeKey<?>, Object> e : attributes.asMap().entrySet()) {
+                if (!pointData.getAttributes().get(e.getKey()).equals(e.getValue())) {
```

Comment:
This will NPE if `e.getKey()` isn't in the attributes.

Reply:
Good point, thank you

Reply:
Pushed a commit to fix this.